src/algorithm/binarySearchTree.ts

```diff
@@ -0,0 +1,113 @@
+class TreeNode<Key, Val> {
+	public key: Key
+	public val: Val
+	public right: TreeNode<Key, Val>
+	public left: TreeNode<Key, Val>
+	constructor(key: Key, val: Val) {
+		this.key = key
+		this.val = val
+	}
+}
+
+export class binarySearchTree<Key, Val> {
+	private root: TreeNode<Key, Val>
+
+	insert(key: Key, val: Val): void {
```

Comment:
looks quite nice and simple, but I have one thing that bug with this the definition, the generic `Key` where at the end this structure requires a sortable scalar type (like number).
```ts
insert(key: Key, val: Val)
```

But something like this doesnt seem to make sense:
```ts
new binarySearchTree<AnotherClass, MyUserClass>()
```

So for me (or maybe I misunderstand things):
* `Key` should be a number (almost every time)
* `Value` can be any data stored by the user and that's what the `Val` type represent

and I kinda expect the tree to be sorted by this `Key` (you are doing it few lines later with `key < x.key`).
But in the unit test, you use ascii character as Key 🤔 

So maybe we should use a predefined mixed type like ```key: number | string``` for this key

And I wonder if we cannot simplify the signature, and implicitly set the key type
```ts
const tree = new binarySearchTree<MyUserClass>()
tree.insert(1, new MyUserClass("tuna"))
tree.insert(4, new MyUserClass("john"))
```

Reply:
As you said, for the most of time `key` would be `number`
and I agree that it doesn't make sense that `key` is some kind of object.  
However I want this to be generic enough to store any keys like Dictionary in C# or map in JS.
So I want to go with mixed type `key: number  | string` as you suggested. 😊 

I will see if I can simply key because of this change.

Reply:
ha ok I see what you wanted to do, it makes sense and yes mixing string and number in the same tree can be a bit annoying.

In this case, we can at least restraint the generic type 😄 

![image](https://user-images.githubusercontent.com/4903529/68304393-020ce180-00e9-11ea-94e3-cfb16218bb9e.png)

TS let you put constraint or interface on generics. So you can do something like that:
```ts
class TreeNode<Key extends number | string, Val> {
   // ...
}

export class binarySearchTree<Key extends number | string, Val> {
   // ...
}
```

Reply:
Oh.. That is cool I will change it to that way 👍